packages/gallery/src/components/helpers/thumbnailsLogic.ts

```diff
@@ -0,0 +1,278 @@
+import { GalleryProps, GALLERY_CONSTS, utils } from 'pro-gallery-lib';
+import { CSSProperties } from 'react';
+
+function calculateActiveIndexOffset({
+  activeIndex,
+  prevActiveIndex,
+  activeIndexOffsetMemory,
+  items,
+}) {
+  const itemsLength = items.length;
+  if (activeIndex === prevActiveIndex) {
+    return activeIndexOffsetMemory;
+  }
+  activeIndex = utils.inRange(activeIndex, itemsLength);
+  const initialRoute = Math.abs(prevActiveIndex - activeIndex);
+  const jumpForwardRoute = Math.abs(
+    prevActiveIndex - itemsLength - activeIndex
+  );
+  const jumpBackwardRoute = Math.abs(
+    prevActiveIndex + itemsLength - activeIndex
+  );
+  if (
+    jumpBackwardRoute < jumpForwardRoute &&
+    jumpBackwardRoute < initialRoute
+  ) {
+    return activeIndexOffsetMemory - itemsLength;
+  }
+  if (jumpForwardRoute < jumpBackwardRoute && jumpForwardRoute < initialRoute) {
+    return activeIndexOffsetMemory + itemsLength;
+  }
+  return activeIndexOffsetMemory;
+}
+
+export function clearGalleryItems(items: any[], galleryItems: any[]): any[] {
+  const clear = (list) =>
+    utils
+      .uniqueBy(list, 'idx')
+      .filter((item) => item.idx !== undefined)
+      .sort((a, b) => a.idx - b.idx);
+  const clearedGalleryItems = clear(galleryItems);
+  return clearedGalleryItems.map((item) => {
+    const realItem = items.find((clearedItem) => clearedItem.id === item.id);
+    return {
+      item: realItem,
+      thumbnailItem: item,
+      idx: item.idx,
+    };
+  });
+}
+
+export function getThumbnailsData({
+  options,
+  activeIndex,
+  items,
+  thumbnailPosition,
+  galleryStructure,
+  galleryWidth,
+  galleryHeight,
+  activeIndexOffsetMemory = activeIndex,
+  prevActiveIndex = activeIndex,
+}: {
+  options: GalleryProps['options'];
+  activeIndex: number;
+  items: GalleryProps['items'];
+  thumbnailPosition: string;
+  galleryStructure: any;
+  galleryWidth: number;
+  galleryHeight: number;
+  activeIndexOffsetMemory?: number;
+  prevActiveIndex?: number;
+}) {
+  const galleryItems = clearGalleryItems(items, galleryStructure.galleryItems);
```

Comment:
[request]
please add a unit test that tests a standard input/output of this function. It's doing a lot and if it will be refactored we would want to know it's not broken without opening a browser